Paco/src/com/pacoapp/paco/sensors/android/RuntimePermissions.java

```diff
@@ -0,0 +1,588 @@
+package com.pacoapp.paco.sensors.android;
+
+import android.accessibilityservice.AccessibilityService;
+import android.annotation.TargetApi;
+import android.content.Context;
+import android.content.Intent;
+import android.content.pm.ApplicationInfo;
+import android.content.pm.PackageManager;
+import android.os.Build;
+import android.os.Bundle;
+import android.util.Log;
+import android.view.accessibility.AccessibilityEvent;
+import android.view.accessibility.AccessibilityNodeInfo;
+
+import com.pacoapp.paco.PacoConstants;
+import com.pacoapp.paco.model.Experiment;
+import com.pacoapp.paco.sensors.android.procmon.EncounteredPermissionRequest;
+import com.pacoapp.paco.sensors.android.procmon.RuntimePermissionsAppUtil;
+import com.pacoapp.paco.shared.model2.InterruptCue;
+import com.pacoapp.paco.shared.util.TimeUtil;
+
+import org.joda.time.DateTime;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * This class handles monitoring of runtime permission changes, by implementing an accessibility
+ * service. Whenever a runtime dialog box pops up, or a permission change is made in the settings,
+ * it fires an intent to the BroadcastTriggerService as any other trigger would.
+ * An alternative would be to regularly poll Android's PackageManager to detect any permission
+ * changes, but this would not allow us to see whether a user 'reaffirmed' a permission e.g. by
+ * denying it when it was first requested.
+ * This class assumes an en_us locale and won't work on other languages. It needs to be extended
+ * if we want it to be used in international studies.
+ * This class should only be used on devices running Android 6.0 and up, since older versions don't
+ * support runtime permissions.
+ * This service is enabled when Paco is granted the Accessibility permission. The
+ * BroadcastTriggerService will make sure that only experiments enabling accessibility logging will
+ * receive accessibility events.
+ */
+@TargetApi(Build.VERSION_CODES.LOLLIPOP) // TODO: update to Marshmallow when project SDK changes
+public class RuntimePermissions extends AccessibilityService {
```

Comment:
For now, let's keep it specific to RuntimePermissions.

Thanks,
Bob

Http://www.pacoapp.com/

> On Aug 5, 2016, at 3:19 PM, Bram Bonné <notifications@github.com> wrote:
> 
> In Paco/src/com/pacoapp/paco/sensors/android/RuntimePermissions.java:
> 
> > + * This class handles monitoring of runtime permission changes, by implementing an accessibility
> > + * service. Whenever a runtime dialog box pops up, or a permission change is made in the settings,
> > + * it fires an intent to the BroadcastTriggerService as any other trigger would.
> > + * An alternative would be to regularly poll Android's PackageManager to detect any permission
> > + * changes, but this would not allow us to see whether a user 'reaffirmed' a permission e.g. by
> > + * denying it when it was first requested.
> > + * This class assumes an en_us locale and won't work on other languages. It needs to be extended
> > + * if we want it to be used in international studies.
> > + * This class should only be used on devices running Android 6.0 and up, since older versions don't
> > + * support runtime permissions.
> > + * This service is enabled when Paco is granted the Accessibility permission. The
> > + * BroadcastTriggerService will make sure that only experiments enabling accessibility logging will
> > + * receive accessibility events.
> > + */
> > +@TargetApi(Build.VERSION_CODES.LOLLIPOP) // TODO: update to Marshmallow when project SDK changes
> > +public class RuntimePermissions extends AccessibilityService {
> Sounds good! If you want, I can change it to AccessibilityEventMonitorService, if you'd like to generalize it to capture different events afterwards? Let me know which one you prefer and I'll change it.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>